File: selenium/pages/base_page.py
from typing import Tuple

from selenium.webdriver.common.keys import Keys
from selenium.webdriver.remote.webdriver import WebDriver
from selenium.webdriver.remote.webelement import WebElement
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import WebDriverWait

Locator = Tuple[str, str]


class BasePage:
    """Shared Selenium page helpers with explicit waits and stable interactions."""

    DEFAULT_TIMEOUT = 15

    def __init__(self, driver: WebDriver, base_url: str) -> None:
        self.driver = driver
        self.base_url = base_url.rstrip("/")
        self.wait = WebDriverWait(driver, self.DEFAULT_TIMEOUT)

    def goto(self, path: str = "/") -> None:
        if path.startswith("http"):
            self.driver.get(path)
        else:
            self.driver.get(f"{self.base_url}{path}")

    def fill_input(self, locator: Locator, value: str) -> None:
        # React inputs can ignore clear(); force value updates + events.
        field = self.wait.until(EC.element_to_be_clickable(locator))
        field.click()
        field.send_keys(Keys.CONTROL, "a", Keys.BACKSPACE)
        if value:
            field.send_keys(value)
        self.driver.execute_script(
            """
            const el = arguments[0];
            const val = arguments[1];
            const setter = Object.getOwnPropertyDescriptor(
                window.HTMLInputElement.prototype, 'value'
            ).set;
            setter.call(el, val);
            el.dispatchEvent(new Event('input', { bubbles: true }));
            el.dispatchEvent(new Event('change', { bubbles: true }));
            el.dispatchEvent(new Event('blur', { bubbles: true }));
            """,
            field,
            value,
        )

    def click_element(self, locator: Locator) -> None:
        # JS click reduces flakiness in headless/overlay-heavy CI environments.
        element = self.wait.until(EC.element_to_be_clickable(locator))
        self.driver.execute_script(
            "arguments[0].scrollIntoView({block: 'center'}); arguments[0].click();",
            element,
        )

File: java/selenium/src/test/java/com/tnd/selenium/pages/BasePage.java
package com.tnd.selenium.pages;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.time.Duration;
import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

/**
 * Base class for Java Selenium page objects.
 *
 * <p>Purpose: centralize wait/click/fill primitives so page classes stay focused on behavior.
 */
public class BasePage {
    private static final String DEBUG_LOG_PATH =
            "/home/technicalnarrativedesigner/Encrypted/Projects/TND/tnd-qa/.cursor/debug-3ba741.log";
    private static final String DEBUG_SESSION_ID = "3ba741";

    protected final WebDriver driver;
    protected final WebDriverWait wait;
    protected final String baseUrl;

    /**
     * Creates reusable Selenium page-object foundation with explicit wait strategy.
     */
    public BasePage(WebDriver driver, String baseUrl) {
        this.driver = driver;
        this.baseUrl = baseUrl;
        this.wait = new WebDriverWait(driver, Duration.ofSeconds(15));
    }

    /**
     * Navigates to either absolute URL or route relative to base URL.
     */
    protected void goTo(String path) {
        if (path.startsWith("http")) {
            driver.get(path);
        } else {
            driver.get(baseUrl + path);
        }
    }

    /**
     * Clicks element using JS after wait/scroll to reduce headless flakiness.
     */
    protected void click(By locator) {
        // JS click keeps interactions stable in headless or scroll-heavy pages.
        WebElement element = wait.until(ExpectedConditions.elementToBeClickable(locator));
        ((JavascriptExecutor) driver).executeScript("arguments[0].scrollIntoView({block: 'center'});", element);
        ((JavascriptExecutor) driver).executeScript("arguments[0].click();", element);
    }

    /**
     * Fills input field after visibility wait.
     */
    protected void fill(By locator, String value) {
        WebElement element = wait.until(ExpectedConditions.visibilityOfElementLocated(locator));
        element.click();
        element.clear();
        element.sendKeys(value);

        // Some CI/headless runs on React-like forms can drop sendKeys updates.
        // If value did not stick, force-set value and dispatch input/change events.
        String currentValue = element.getAttribute("value");
        if (currentValue == null || !currentValue.equals(value)) {
            ((JavascriptExecutor) driver).executeScript(
                    "arguments[0].value = arguments[1];"
                            + "arguments[0].dispatchEvent(new Event('input', { bubbles: true }));"
                            + "arguments[0].dispatchEvent(new Event('change', { bubbles: true }));",
                    element,
                    value
            );
        }
    }

    /**
     * Writes one NDJSON debug event for runtime hypothesis validation.
     */
    protected void debugLog(String runId, String hypothesisId, String location, String message, String dataJson) {
        String line = "{\"sessionId\":\""
                + jsonEscape(DEBUG_SESSION_ID)
                + "\",\"runId\":\""
                + jsonEscape(runId)
                + "\",\"hypothesisId\":\""
                + jsonEscape(hypothesisId)
                + "\",\"location\":\""
                + jsonEscape(location)
                + "\",\"message\":\""
                + jsonEscape(message)
                + "\",\"data\":"
                + dataJson
                + ",\"timestamp\":"
                + System.currentTimeMillis()
                + "}";
        try {
            Path logPath = Path.of(DEBUG_LOG_PATH);
            if (logPath.getParent() != null) {
                Files.createDirectories(logPath.getParent());
            }
            Files.writeString(
                    logPath,
                    line + System.lineSeparator(),
                    StandardOpenOption.CREATE,
                    StandardOpenOption.APPEND
            );
        } catch (IOException ignored) {
            // Never fail the test because of debug logging.
        }
        System.out.println(line);
    }

    /**
     * Escapes string content for JSON literals in debug events.
     */
    protected String jsonEscape(String value) {
        if (value == null) {
            return "";
        }
        return value
                .replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r");
    }
}
